ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java

```diff
@@ -540,6 +549,43 @@ private EndpointResponse handleScalablePushQuery(
     return EndpointResponse.ok(queryStreamWriter);
   }
 
+  private EndpointResponse handleStreamPullQuery(
+      final ImmutableAnalysis analysis,
+      final ServiceContext serviceContext,
+      final ConfiguredStatement<Query> configured,
+      final CompletableFuture<Void> connectionClosedFuture) {
+
+    final TransientQueryMetadata transientQueryMetadata = ksqlEngine
+        .createStreamPullQuery(
+            serviceContext,
+            configured,
+            false
+        );
+    localCommands.ifPresent(lc -> lc.write(transientQueryMetadata));
+
+    // Be sure to close the stream in the event of any error. It's safe to close it twice,
+    // which will happen if we call close inside the block, and then again when we exit the
+    // try-with-resources
+    try (QueryStreamWriter queryStreamWriter = new QueryStreamWriter(
+        transientQueryMetadata,
+        disconnectCheckInterval.toMillis(),
+        OBJECT_MAPPER,
+        connectionClosedFuture
+    )) {
+      ksqlEngine.waitForStreamPullQuery(
+          serviceContext,
+          analysis,
+          configured,
+          transientQueryMetadata
+      );
+      // stop the response stream before sending the successful response.
+      queryStreamWriter.close();
+      return EndpointResponse.ok(queryStreamWriter);
+    } catch (final KsqlServerException e) {
+      throw new KsqlApiException(e.getMessage(), HttpResponseStatus.INTERNAL_SERVER_ERROR.code());
```

Comment:
I'm confused why this gets wrapped in a KsqlApiException.  I assume it's so that we can return a 500 error rather than a 400, though `KsqlApiException` seems to be intended for 400 errors.  We're so inconsistent in all of our broader code about which exceptions we throw and how they map to error codes, I cannot recommend what to do here.  We should prioritize making all query related endpoints consistent @agavra .

Reply:
Interesting. I thought it was just intended to be able to specify the error code, from the constructor. Is there another exception type for 5xx ?

Reply:
Ah, upon further investigation, the exception handling does need to be handled more broadly. Apparently, there's insufficient test coverage, and it actually looks like the HTTP/1 API turns all exceptions into 400s regardless of type.

I'll schedule a follow-on PR to address that.

Reply:
Yep, we should prioritize that for next release.